algorithms task (5)/student/include/impl.h

```diff
@@ -10,4 +13,19 @@
  * @return gathered range
  */
 template <typename BiIt, typename UnaryPredicate>
-std::pair<BiIt, BiIt> gather(BiIt begin, BiIt end, BiIt position, UnaryPredicate predicate);
+std::pair<BiIt, BiIt> gather(BiIt begin, BiIt end, BiIt position, UnaryPredicate predicate)
+{
+    const auto notPredicate = [predicate](const typename std::iterator_traits<BiIt>::reference value)
+    {
+        return !predicate(value);
+    };
+
+    auto partitionBegin = std::partition(begin, std::next(position), notPredicate);
+    auto partitionEnd = std::partition(position, end, predicate);
+
+    if (partitionBegin == std::next(position))
+    {
+        std::advance(partitionBegin, -1);
```

Comment:
What if partition begin is at "begin"?

Reply:
Partition's begin cannot be "begin" here as it equals to std::next(position), so it will be at least "begin" + 1.

Reply:
please add tests for empty range

Reply:
Right! Nothing is added, the test for empty range already exists. But the algorithm is changed.